lib/batch_verification/src/verification_client.rs

```diff
@@ -0,0 +1,306 @@
+}
+
+#[derive(Debug, thiserror::Error)]
+enum BatchVerificationError {
+    #[error("Missing records for block {0}")]
+    MissingBlock(u64),
+    #[error("Tree error")]
+    TreeError,
+    #[error("Batch data mismatch: {0}")]
+    BatchDataMismatch(String),
+}
+
+impl BatchVerificationClient {
+    pub fn new(
+        private_key: SecretString,
+        chain_id: u64,
+        diamond_proxy: Address,
+        server_address: String,
+    ) -> Self {
+        Self {
+            signer: PrivateKeySigner::from_str(private_key.expose_secret())
+                .expect("Invalid batch verification private key"),
+            chain_id,
+            diamond_proxy,
+            block_storage: HashMap::new(),
+            server_address,
+        }
+    }
+
+    async fn handle_verification_request(
+        &self,
+        request: BatchVerificationRequest,
+    ) -> Result<BatchSignature, BatchVerificationError> {
+        tracing::info!(
+            "Handling batch verification request {} for batch {} (blocks {}-{})",
+            request.request_id,
+            request.batch_number,
+            request.first_block_number,
+            request.last_block_number,
+        );
+
+        let blocks: Vec<(&BlockOutput, &ReplayRecord, TreeBatchOutput)> =
+            (request.first_block_number..=request.last_block_number)
+                .map(|block_number| {
+                    let (block_output, replay_record, tree_data) = self
+                        .block_storage
+                        .get(&block_number)
+                        .ok_or(BatchVerificationError::MissingBlock(block_number))?;
+
+                    let (root_hash, leaf_count) = tree_data
+                        .block_start
+                        .clone()
+                        .root_info()
+                        .map_err(|_| BatchVerificationError::TreeError)?;
+
+                    let tree_output = TreeBatchOutput {
+                        root_hash,
+                        leaf_count,
+                    };
+                    Ok((block_output, replay_record, tree_output))
+                })
+                .collect::<Result<Vec<_>, BatchVerificationError>>()?;
+
+        // TODO VALIDATE
+        let commit_batch_info = CommitBatchInfo::new(
+            blocks
+                .iter()
+                .map(|(block_output, replay_record, tree)| {
+                    (
+                        *block_output,
+                        &replay_record.block_context,
+                        replay_record.transactions.as_slice(),
+                        tree,
+                    )
+                })
+                .collect(),
+            self.chain_id,
+            self.diamond_proxy,
+            request.batch_number,
+        )
+        .into_l1_commit_data(BatchDaInputMode::Rollup);
+
+        if commit_batch_info != request.commit_data {
+            let mut mismatches = Vec::new();
+
+            // I don't like this, but it works great :/
```

Comment:
Are you doing this because you want a pretty-printed diff between these two structs?

FWIW there is [structdiff](https://github.com/knickish/structdiff) that I have used once in the past. You should be able to derive it for `CommitBatchInfo` pretty easily. Another reason to use it over `CommitBatchInfoZKsyncOS` from alloy_sol :wink: 

@popzxc maybe you have a better suggestion?